### ISSUE TITLE
Finalize Conversion/Transaction Format

### ISSUE THREAD

**Issue by WhoSoup** [pegnet/pegnet#283]
This step is critical for all further work. @sambarnes has been working on https://hackmd.io/EfSzduXqQ8yyD9vBYK3E-w?edit which is the current up-to-date spec of the entry's content. The extids derive from the FAT-0 spec

We need to:
* Determine if FAT-0 extids are enough or we need additional information in the extids
* Finalize the structure of the content to fit our current needs and be adaptable to our future needs
* Conform the format to any design philosophies required by FAT (such as naming conventions)
* Update the [FATIP](https://github.com/Factom-Asset-Tokens/FAT/pull/25) with the final format 

There was still some discussion as to which parameters are optional, such as conversions not requiring an output. If you see anything lacking, please mention it.

**Comment by sambarnes** [comment]
Just pushed current work to my own fatd fork here: https://github.com/sambarnes/fatd/tree/fat2/fat/fat2

Tackling pure transactions first, postponing conversion support until #293 is completed and we have the conversion math code completed and able to be dropped in

**Comment by PaulBernier** [comment]
While working on one of my own use case of FAT transaction/conversion, I realized that we could use a very common trick with great benefit here: allowing the content of the factom entries to be in a compressed format. Compression of JSON is extremely efficient when you have arrays and repeated fields, which is exactly the case of FAT transactions. 

**Suggestion:** allow the JSON stored in the entry to be compressed with a determined. Deflate or gzip are probably the best candidates due to their wide availability (and they are the compression schemes of HTTP).

**Benefit:** for batch transactions, or multi output transactions the gain of space (and so the reduced cost in EC) can be tremendous (for my specific use case I get 65% to 75% reduction of space, more below)

**Impact:** the implementation has to try to parse a JSON, or try to decompress the entry first. This can have some performance hit. But the scope of that change is contained to just the place where the entry is parse, after that the structure remain the same, which should have a minimal impact on the existing flow of execution.

## Some results

Here's a Node.js script I wrote to evaluate the gain in term of space: https://pastebin.com/ZEkZtQ8d
Note that I am using randomly generated amounts and FCT addresses for the outputs, if anything repeat, the gain is even bigger.

Feel free to tweak the script to experiment other scenarios.

### With just 10 tx batched

- JSON: https://pastebin.com/cG8yLPgK
- Size: 2034
- gzip compression: 667 bytes, 32% of the uncompressed size
- deflate compression: 655 bytes, 32% of the uncompressed size
- brotli compression: 647 bytes, 31.8% of the uncompressed size

### With just 100 tx batched

- JSON: https://pastebin.com/AJ5aG706
- Size: 20377
- gzip compression: 5166 bytes, 25.3% of the uncompressed size
- deflate compression: 5154 bytes, 25.2% of the uncompressed size
- brotli compression: 4857 bytes, 23.8% of the uncompressed size

**Comment by Emyrk** [comment]
@PaulBernier I like the idea of compression, but I fear attacks like zip bombs being places on-chain. Is there any way to limit the uncompressed size so we can avoid maliciously zipped contents to either explode or be computationally expensive?

**Comment by PaulBernier** [comment]
Good point @Emyrk it needs more thoughts overall anyway, but the gain of the compression were significant enough for me to bring the topic on the table. This probably needs its own separate discussion.

**Comment by WhoSoup** [comment]
If you want to fix JSON's size bloat the best way to do it would be to not use JSON and instead use something else (like protobufs :smirk_cat: ) imo

**Comment by PaulBernier** [comment]
Right, I was thinking that compressing JSON would be an easy to implement intermediary solution, but I agree that it is a compromise and that the root issue should probably be addressed thoroughly in a better way.
